p2p/peer.go

```diff
@@ -214,97 +214,280 @@ type PeerUpdate struct {
+//
+// NOTE: We can't take an address here, since e.g. TCP uses a different port
+// number for outbound traffic than inbound traffic, so the peer's endpoint
+// wouldn't necessarily be an appropriate address to dial.
+func (m *peerManager) Accepted(peerID NodeID) error {
+	m.mtx.Lock()
+	defer m.mtx.Unlock()
+	peer, err := m.store.Get(peerID)
+	if err != nil {
+		return err
+	} else if peer == nil {
+		peer = newPeerInfo(peerID)
+		if err = m.store.Set(peer); err != nil {
+			return err
+		}
+	}
+	if m.connected[peerID] {
+		return fmt.Errorf("peer %q is already connected", peerID)
 	}
-	peer.AddAddress(address)
+	m.connected[peerID] = true
 	return nil
 }
 
-// Claim claims a peer. The caller has exclusive ownership of the peer, and must
-// return it by calling Return(). Returns nil if the peer could not be claimed.
-// If the peer is not known to the store, it is registered and claimed.
-func (s *peerStore) Claim(id NodeID) *peerInfo {
-	s.mtx.Lock()
-	defer s.mtx.Unlock()
-	if s.claimed[id] {
-		return nil
+// Ready marks a peer as ready, broadcasting status updates to subscribers. The
+// peer must already be marked as connected. This is separate from Dialed() and
+// Accepted() to allow the router to set up its internal queues before reactors
+// start sending messages (holding the Router.peerMtx mutex while calling
+// Accepted or Dialed will halt all message routing while peers are set up, which
+// is too expensive and also causes difficulties in tests where we may want to
+// consume peer updates and send messages sequentially).
+//
+// FIXME: This possibly indicates an architectural problem. Should the peerManager
+// handle actual network connections to/from peers as well? Or should all of this
+// be done by the router?
+func (m *peerManager) Ready(peerID NodeID) {
+	m.mtx.Lock()
+	connected := m.connected[peerID]
+	m.mtx.Unlock()
+	if connected {
+		m.broadcast(PeerUpdate{
+			PeerID: peerID,
+			Status: PeerStatusUp,
+		})
+	}
+}
+
+// Disconnected unmarks a peer as connected, allowing new connections to be
+// established.
+func (m *peerManager) Disconnected(peerID NodeID) error {
+	m.mtx.Lock()
+	defer m.mtx.Unlock()
+	delete(m.connected, peerID)
+	m.broadcast(PeerUpdate{
+		PeerID: peerID,
+		Status: PeerStatusDown,
+	})
+	return nil
+}
+
+// peerStore stores information about peers. It is currently a bare-bones
+// in-memory store, and will be fleshed out later.
+//
+// peerStore is not thread-safe, since it assumes it is only used by peerManager
+// which handles concurrency control. This allows multiple operations to be
+// executed atomically, since the peerManager will hold a mutex while executing.
+type peerStore struct {
+	peers map[NodeID]peerInfo
+}
+
+// newPeerStore creates a new peer store.
+func newPeerStore() *peerStore {
+	return &peerStore{
+		peers: map[NodeID]peerInfo{},
 	}
+}
+
+// Get fetches a peer, returning nil if not found.
+func (s *peerStore) Get(id NodeID) (*peerInfo, error) {
```

Comment:
This never returns an error. Why not just `(peerInfo, bool)`?

Reply:
It will be fallible once `peerStore` is backed by an actual database (which can have IO errors).